runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,320 @@
+        bufstates[bufnr] = bufstate
+        if api.nvim_buf_is_loaded(bufnr) then
+          foldupdate(bufnr)
+        end
+      end
+    end,
+  })
+  api.nvim_create_autocmd('LspNotify', {
+    group = augroup_refresh,
+    buffer = bufnr,
+    callback = function(args)
+      local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+      if
+        client:supports_method(ms.textDocument_foldingRange, bufnr)
+        and (
+          args.data.method == ms.textDocument_didChange
+          or args.data.method == ms.textDocument_didOpen
+        )
+      then
+        refresh(bufnr, args.data.client_id)
+      end
+    end,
+  })
+
+  for _, client in
+    ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+  do
+    refresh(bufnr, client.id)
+  end
+
+  return bufstate
+end
+
+---@param winid integer
+---@param ranges lsp.FoldingRange[]
+local function foldclose(winid, ranges)
+  vim._with({ win = winid }, function()
+    -- Ensuring folding the smallest ranges first is not explicitly supported,
+    -- but since LSP servers often return the largest ranges first,
+    -- we traverse the ranges in reverse order here.
+    --
+    -- If sorting needs to be implemented,
+    -- the most suitable algorithm here is
+    -- [Timesort](https://wikipedia.org/wiki/Timsort).
+    --
+    -- "Hanc marginis exiguitas non caperet."
+    for i = #ranges, 1, -1 do
+      local range = ranges[i]
+      local start_lnum = range.startLine + 1
+      local end_lnum = range.endLine + 1
+      -- Folding a single line is not supported.
+      if start_lnum ~= end_lnum then
+        vim.cmd(start_lnum .. ',' .. end_lnum .. 'foldclose')
+      end
+    end
+  end)
+end
+
+--- Close all {kind} of folds in the the window with {winid}.
+---
+--- To automatically fold imports when opening a file, you can use an autocmd:
+---
+--- ```lua
+--- vim.api.nvim_create_autocmd('LspNotify', {
+---   callback = function(args)
+---     if args.data.method == 'textDocument/didOpen' then
+---       vim.lsp.folding_range.foldclose('imports', vim.fn.bufwinid(args.buf))
+---     end
+---   end,
+--- })
+--- ```
+---
+---@param kind lsp.FoldingRangeKind "comment" | "imports" | "region"
+---@param winid? integer Defaults to the current window.
+function M.foldclose(kind, winid)
+  vim.validate('kind', kind, 'string')
+  vim.validate('winid', winid, 'number', true)
+
+  winid = winid or api.nvim_get_current_win()
+  local bufnr = api.nvim_win_get_buf(winid)
+  local bufstate = bufstates[bufnr]
+  if not bufstate then
+    return
+  end
+  ---@cast bufstate vim.lsp.folding_range.BufState
+
+  -- Defer `foldclose()` if the buffer is not up-to-date.
+  if bufstate.version ~= util.buf_versions[bufnr] then
+    for _, client in
+      ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+    do
+      local request_id = refresh(bufnr, client.id)
+      api.nvim_create_autocmd('LspRequest', {
```

Comment:
This autocmd seems a bit redundant, isn't the callback unconditional?

Reply:
This is to avoid executing `foldclose` when the buffer has already been modified, but the bufstate has not yet been updated. This ensures that the result of `foldclose` corresponds to the modified buffer.

More specifically, there is a situation where `foldclose` is executed immediately when opening a file (common in autocmd, such as the automatic folding import example given above). At this point, the bufstate is likely not updated yet, but what you actually want is the result after it has been updated, corresponding to the content you see in the buffer.

Reply:
So you basically want to defer this code until all lines in a redraw have been calculated?

Reply:
It is not for redrawing, but to wait for the `foldingRange` request to get a result.

Reply:
Oh, because `LspRequest` is triggered on the handler. Ok can you just add some comments explaining this. I'd probably want to re-implement this in the future when some other things have been added.